File: imagepipeline-base/src/main/java/com/facebook/imagepipeline/cache/BoundedLinkedHashSet.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.cache;

import com.facebook.infer.annotation.Nullsafe;
import java.util.LinkedHashSet;
import javax.annotation.concurrent.ThreadSafe;

@ThreadSafe
@Nullsafe(Nullsafe.Mode.STRICT)
public class BoundedLinkedHashSet<E> {

  private int mMaxSize;
  private LinkedHashSet<E> mLinkedHashSet;

  public BoundedLinkedHashSet(final int maxSize) {
    mLinkedHashSet = new LinkedHashSet<>(maxSize);
    mMaxSize = maxSize;
  }

  public synchronized boolean contains(E o) {
    return mLinkedHashSet.contains(o);
  }

  public synchronized boolean add(E key) {
    if (mLinkedHashSet.size() == mMaxSize) {
      mLinkedHashSet.remove(mLinkedHashSet.iterator().next());
    }
    mLinkedHashSet.remove(key);
    return mLinkedHashSet.add(key);
  }
}

File: imagepipeline-base/src/main/java/com/facebook/imagepipeline/cache/BoundedLinkedHashSet.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.cache

import java.util.LinkedHashSet
import javax.annotation.concurrent.ThreadSafe

@ThreadSafe
class BoundedLinkedHashSet<E>(maxSize: Int) {

  private val maxSize: Int = maxSize
  private val linkedHashSet: LinkedHashSet<E> = LinkedHashSet(maxSize)

  @Synchronized operator fun contains(o: E): Boolean = linkedHashSet.contains(o)

  @Synchronized
  fun add(key: E): Boolean {
    if (linkedHashSet.size == this.maxSize) {
      linkedHashSet.remove(linkedHashSet.iterator().next())
    }
    linkedHashSet.remove(key)
    return linkedHashSet.add(key)
  }
}
